### PR TITLE
Turn off bed while probing during MBL

### PR DESCRIPTION
**Pull request by leptun**
Many people have been experiencing retries in the probing during Mesh Bed Leveling after the 7x7 version was introduced. It has been determined that the bed heating was causing interference to the Pinda probe and it would cause false triggers. This PR disables the heating right when the probing happens so that no switch of state (on->off or off->on). This greatly reduces the retries count.
Overall, the probing accuracy is not increased, it's just the retries that are mitigated.
Issue describing the issue further:
Closing #2280 
Closing #1651 
@DRracer 

PFW-1019

### CODE REVIEW

File: Firmware/heatbed_pwm.cpp [hunk 1]
```diff
@@ -137,12 +140,17 @@ ISR(TIMER0_OVF_vect)          // timer compare interrupt service routine
 	case States::ONE:             // state ONE - we'll either stay in ONE or change to FALL
 		OCR0B = 255;
 		slowCounter += slowInc;   // this does software timer_clk/256 or less
-		if( slowCounter < pwm ){
-			return;
+		if (!bedPWMDisabled){ //disable heating as soon as possible
```
Comment (DRracer) [inline Firmware/heatbed_pwm.cpp]: is it really correct to have the slowCounter outside of the "if" statement? It basically continues to tick the SW (aka slow) PWM. Wouldn't it be better to just keep the States::ONE when bed PWM is to be disabled?

File: Firmware/heatbed_pwm.cpp [hunk 2]
```diff
@@ -137,12 +140,17 @@ ISR(TIMER0_OVF_vect)          // timer compare interrupt service routine
...
+			}
+			if( (soft_pwm_bed << 1) >= (255 - slowInc - 1) ){  //@@TODO simplify & explain
+				// if slowInc==2, soft_pwm == 251 will be the first to do short drops to zero. 252 will keep full heating
+				return;           // want full duty for the next ONE cycle again - so keep on heating and just wait for the next timer ovf
+			}
 		}
-		if( (soft_pwm_bed << 1) >= (255 - slowInc - 1) ){  //@@TODO simplify & explain
-			// if slowInc==2, soft_pwm == 251 will be the first to do short drops to zero. 252 will keep full heating
-			return;           // want full duty for the next ONE cycle again - so keep on heating and just wait for the next timer ovf
+		else if (pwm > 200){ //if duty cycle is high and BED PWM is disabled keep heater on. Prevents overcooling
```
Comment (DRracer) [inline Firmware/heatbed_pwm.cpp]: is this "else" really necessary (with the above comment in mind)?

File: Firmware/Marlin_main.cpp [hunk 3]
```diff
@@ -4973,11 +4973,7 @@ if(eSoundMode!=e_SOUND_MODE_SILENT)
 					z_offset_u = eeprom_read_word((uint16_t*)(EEPROM_BED_CALIBRATION_Z_JITTER + 2 * (ix + iy * 3 - 1)));
 				}
 				z0 = mbl.z_values[0][0] + *reinterpret_cast<int16_t*>(&z_offset_u) * 0.01;
-				#ifdef SUPPORT_VERBOSITY
-				if (verbosity_level >= 1) {
-					printf_P(PSTR("Bed leveling, point: %d, calibration Z stored in eeprom: %d, calibration z: %f \n"), mesh_point, z_offset_u, z0);
-				}
-				#endif // SUPPORT_VERBOSITY
+				// printf_P(PSTR("Bed leveling, point: %d, calibration Z stored in eeprom: %d, calibration z: %f \n"), mesh_point, z_offset_u, z0);
```
Comment (DRracer) [inline Firmware/Marlin_main.cpp]: probably we can remove this change, it is not related to the features in this PR

Comment (leptun) [reply Firmware/Marlin_main.cpp]: Reverted the testing change

File: Firmware/heatbed_pwm.cpp [hunk 4]
```diff
@@ -45,6 +45,12 @@
 // If there are any change requirements in the future, the signal must be checked with an osciloscope again,
 // ad-hoc changes may completely screw things up!
 
+// 2020-01-29 update: we are introducing a new option to the automaton that will allow us to force the output state
+// to either full ON or OFF. This is so that interference during the MBL probing is minimal.
+// To accomplish this goal we use bedPWMDisabled. It is only supposed to be used for brief periods of time as to
+// not make the bed temperature too unstable. Also, careful consideration should be used when using this
+// option as leaving this enabled will also keep the bed output in the state it stopped in.
+
```
Comment (DRracer) [inline Firmware/heatbed_pwm.cpp:53]: excellent explanation

**Review by DRracer — APPROVED** [review]

**Comment by GilesBathgate** [comment]
Hi @leptun I wanted to try this out on MK2 firmware. I know it was to solve an issue with @DRracer's fast PWM (i.e `heatbed_pwm.cpp`) which only exists in MK3.

However, in my tests with a bed temp heated to only 36°C, I've noticed a considerable difference in probe values if the bed is ON vs OFF and wanted to test this patch.

I've attempted to backport your change here: https://github.com/GilesBathgate/Prusa-Firmware/commit/c70d00aef091025f85508f92e3ff380e316470b5

I wondered if you can take a look and see if I've understood this correctly.

**Comment by leptun** [comment]
@GilesBathgate Hello. Yes, this is what I had in mind when I implemented it here. One thing to note, though: With the heatbed_pwm.cpp the idea was to avoid the fast PWM pulses when switching the state. The way you implemented it follows exactly what we did, but it might not be that useful overall. You can try and see if you get improfements with it as it is now, but you might want to try something like this:
Put the `if (!bedPWMDisabled) {...}` so that it only prevents a WRITE(1) to the bed. This way it will always stay off during the probing. If you use it as it is now the bed will stay either on or off (no switching)

**Comment by GilesBathgate** [comment]
@leptun So leave the state changing code in place and just disable the writes? 

i.e `if (!bedPWMDisabled) WRITE(HEATER_BED_PIN, 1);`
and `if (!bedPWMDisabled) WRITE(HEATER_BED_PIN, 0);`

**Comment by GilesBathgate** [comment]
@leptun So you suggest only disabling the write that turns the bed on i.e:

`if (!bedPWMDisabled) WRITE(HEATER_BED_PIN, 1);`

but leave all the other code to execute.

**Comment by leptun** [comment]
Yes, that should work. So that the bed will disable itself once the PWM cycle finishes, but it won't get enabled again. Not sure what happens when the bed is at full power, but otherwise it should be fine

**Comment by GilesBathgate** [comment]
>Not sure what happens when the bed is at full power.

It gets hot! :joy:

**Comment by GilesBathgate** [comment]
@leptun So to test whether this was working I added the following gcode commands:

```c++
    case 88: //just for test
        bedPWMDisabled=1;
        break;
    case 89:
        bedPWMDisabled=0;
        break;
```

I then use the application EMF Detector for my mobile phone to do basic detection of the magnetic flux caused by the PWM. I just place my phone directly on the heated bed.

image

Then I turn the heat bed up to 40°C

image

Unfortunately, I am not able to get the EMF to go back down by issuing my test gcodes.
If I set the temp back to 0°C (I.e OFF) the EMF goes down immediately.

**Comment by GilesBathgate** [comment]
@leptun I figured it out, there was an additional place where `WRITE(HEATED_BED_PIN,1)` was being called: See https://github.com/GilesBathgate/Prusa-Firmware/commit/ad98d7c58fbc72fe824e88894c597fc4001f33db#diff-719af9fc6a59d745b8e9c0164f8dcebfR1532

Works as expected now but I will share some results from actual MBL probing tests next.

**Pull request merged**